internal/provider/waypoint/data_source_waypoint_action.go

```diff
@@ -0,0 +1,207 @@
+								Description: "Key value headers to send with the request.",
+								Computed:    true,
+								ElementType: types.StringType,
+							},
+							"url": schema.StringAttribute{
+								Description: "The full URL this request should make when invoked.",
+								Computed:    true,
+							},
+							"body": schema.StringAttribute{
+								Description: "The body to be submitted with the request.",
+								Computed:    true,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *DataSourceAction) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	d.client = client
+}
+
+func (d *DataSourceAction) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data ActionResourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	client := d.client
+	if d.client == nil {
+		resp.Diagnostics.AddError(
+			"Unconfigured HCP Client",
+			"Expected configured HCP client. Please report this issue to the provider developers.",
+		)
+		return
+	}
+
+	projectID := client.Config.ProjectID
+	if !data.ProjectID.IsNull() {
+		projectID = data.ProjectID.ValueString()
+	}
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      projectID,
+	}
+
+	var actionModel *waypoint_models.HashicorpCloudWaypointActionConfig
+	var err error
+
+	actionModel, err = clients.GetAction(ctx, client, loc, data.ID.ValueString(), data.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(err.Error(), "Failed to find action by ID or name")
+		return
+	}
+
+	if actionModel.ID != "" {
+		data.ID = types.StringValue(actionModel.ID)
+	}
+	if actionModel.Name != "" {
+		data.Name = types.StringValue(actionModel.Name)
+	}
+	if actionModel.Description != "" {
+		data.Description = types.StringValue(actionModel.Description)
+	}
+
+	data.OrgID = types.StringValue(client.Config.OrganizationID)
+	data.ProjectID = types.StringValue(client.Config.ProjectID)
+
+	data.Request = &actionRequest{}
+	headerMap := make(map[string]string)
+
+	var diags diag.Diagnostics
+
+	// In the future, expand this to accommodate other types of requests
+
+	data.Request.Custom = &customRequest{}
+	if actionModel.Request.Custom.Method != nil {
+		methodString, err := convertMethodToStringType(*actionModel.Request.Custom.Method)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unexpected HTTP Method",
+				"Expected GET, POST, PUT, DELETE, or PATCH. Please report this issue to the provider developers.",
+			)
+			return
+		} else {
+			data.Request.Custom.Method = methodString
+		}
```

Comment:
style nit: because the above `if er != nil` always ends in `return`, we can omit the `else` here. It makes the code easier to reason about _in general_, even if in this specific case the `else` branch is just 1 line

Reply:
Specifically meaning this instead:

```golang
if err != nil {
  // stuff
  return
}
data.Request.Custom.Method = methodString
```